controller/src/controller.rs

```diff
@@ -319,15 +348,120 @@ impl<T: BottlerocketShadowClient> BrupopController<T> {
+                .map(|brs| brs.name_any())
+                .collect();
+            let nodes = self.all_nodes();
+            self.bottlerocketshadows_cleanup(nodes, brss_name).await?;
+
+            // Emit metrics at the end of the loop in case the loop didn't progress any nodes.
+            self.emit_metrics()?;
+
+            // Sleep until it's time to check for more action.
+            sleep(ACTION_INTERVAL).await;
+        }
+    }
+
+    /// Runs the event loop for the Brupop controller.
+    ///
+    /// Because the controller wants to gate the number of simultaneously updating nodes, we can't allow the update state machine
+    /// of each individual bottlerocket node to run concurrently and in an event-driven fashion, as is typically done with controllers.
+    /// Instead, we will keep an updated store of `BottlerocketShadow` objects based on cluster events, and then periodically make
+    /// scheduling decisions based on that store.
+    ///
+    /// The controller is designed to run on a single node in the cluster and rely on the scheduler to ensure there is always one
+    /// running; however, it could be expanded using leader-election and multiple nodes if the scheduler proves to be problematic.
+    pub async fn run(&self) -> Result<()> {
+        // generate brupop cron expression schedule
+        let brupop_scheduler = BrupopCronScheduler::new()
+            .context(controllerclient_error::GetCronScheduleSnafu)?
+            .generate_schedule()
+            .await
+            .context(controllerclient_error::GetCronScheduleSnafu)?;
+
+        // TODO: this will be removed when we remove update time window settings
+        let brupop_schedule_default = Schedule::from_str(SCHEDULER_DEFAULT).context(
+            controllerclient_error::GenerateScheduleFailedSnafu {
+                cron: SCHEDULER_DEFAULT,
+            },
+        )?;
+
+        // generate brupop update time window schedule
+        // TODO: this will be removed when we remove update time window settings
+        let update_window_start = read_env_var(UPDATE_WINDOW_START_ENV_VAR)?;
+        let update_window_stop = read_env_var(UPDATE_WINDOW_STOP_ENV_VAR)?;
+
+        // When users still set up time window settings and not use new setting cron expression scheduler, we should
+        // run controller with old setting update time window. Otherwise, we should always prioritize to use cron expression scheduler.
+        // TODO: this will be removed when we remove update time window settings
+        if brupop_scheduler.eq(&brupop_schedule_default) {
+            if update_window_start != UPDATE_WINDOW_START_DEFAULT.to_string()
+                || update_window_stop != UPDATE_WINDOW_STOP_DEFAULT.to_string()
+            {
+                self.run_with_update_time_window(&update_window_start, &update_window_stop)
+                    .await?
+            }
+        } else {
+            self.run_with_cron_scheduler(&brupop_scheduler).await?
+        }
```

Comment:
In general I'm not a fan of having two implementations of `run` which are more or less the same, except for the scheduler implementation. This leaves us in a scenario where we could introduce a bug in one branch of behavior but not the other.

For scenarios like this, I would really prefer to use something like the strategy pattern if possible -- if we had a `BrupopScheduler` trait that helped us decide how to proceed with the updates, I think this would be a lot cleaner. One way you could do this is by having the `Scheduler` use some inputs and its own internal state to return an enum telling Brupop controller what to do, e.g.:

* `CompletePendingUpdates`
* `ProceedWithUpdatingHosts`
* `RefrainFromUpdatingHosts`

Probably a more thoughtful enum could be used, but hopefully this gets the idea across.



Since we're hoping to remove the Window implementation eventually, we could also otherwise share code by trying to move the primary controller logic into shared functions.

Reply:
I'd be open to merging in spite of this feedback, but if we can improve this relatively quickly I'd prefer to.